### PR TITLE
Add rpc complete interface

### LINKED ISSUE #11880: train因Deadline Exceeded连不上ps会core dump，程序应该平稳退出
End pass 4, train_loss 2.88446474075, train_acc1 0.378594011068, train_acc5 0.633796095848, test_loss 2.4196540374, test_acc1 0.44827724359, test_acc5 0.70953525641
F0628 11:17:58.649018   220 grpc_client.cc:248] var: name:[batch_norm_22.w_0] ep:[192.168.16.34:5002] grpc error:Deadline Exceeded
*** Check failure stack trace: ***
    @     0x7f1a1031d07d  google::LogMessage::Fail()
    @     0x7f1a10320b2c  google::LogMessage::SendToLog()
    @     0x7f1a1031cba3  google::LogMessage::Flush()
    @     0x7f1a1032203e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f1a100d0db0  paddle::operators::distributed::GRPCClient::Proceed()
    @     0x7f1a21fafc80  (unknown)
    @     0x7f1a6312f6ba  start_thread
    @     0x7f1a62e6541d  clone
    @              (nil)  (unknown)
Aborted (core dumped)
程序应该平稳退出

### PR DESCRIPTION
**Pull request by Yancey0623**
fixed #11880

### CODE REVIEW

**Review by typhoonzero — COMMENTED** [review]
can you please add a test to test that if 2 trainers have different number of samples they don't hung or timeout.

**Review by typhoonzero — DISMISSED** [review]
LGTM, thanks, the unit test can add in another PR

**Comment by typhoonzero** [comment]
need to update `paddle/fluid/API.spec` to change those API.

**Comment by Yancey0623** [comment]
Can @panyx0718 @typhoonzero help to review the API changed?

**Review by typhoonzero — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/executor.h [hunk 1]
```diff
@@ -44,17 +44,11 @@ class Executor {
...
+   * Calling this method will send complete messages to all pserver instances.
    */
-  void BeginPass();
-
-  /*
-   * Sending signal to pserver to mark current pass finished.
-   */
-  void EndPass();
-#endif
+  void Close();
```
Comment (typhoonzero) [inline paddle/fluid/framework/executor.h:51]: Why changed to `Close`? The underlying method is still called `Complete`

File: paddle/fluid/operators/distributed/grpc_client.cc [hunk 2]
```diff
@@ -35,20 +35,16 @@ void GRPCClient::InitEventLoop() {
...
-  }
-  this->Wait();
-}
-
-void GRPCClient::SendEndPass() {
-  for (auto& it : channels_) {
-    VLOG(3) << "send end pass to " << it.first;
-    this->AsyncSendEndPass(it.first);
+void GRPCClient::SendComplete() {
+  std::unique_lock<std::mutex> lk(completed_mutex_);
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/grpc_client.cc:40]: this lock scope is a bit large.

**Comment by typhoonzero** [comment]
See https://github.com/PaddlePaddle/Paddle/issues/12163, we may need to dedup client request at server side to be a better implement.

File: paddle/fluid/framework/executor.h (hunk 1 above)
Comment (panyx0718) [reply paddle/fluid/framework/executor.h:51]: complete is a subsetof close(). close can also clean up other resources in the future. It's similar to tf's ```Session.close```. It cleans up all related resources (such as complelte) and shutdown the session.

**Comment by panyx0718** [comment]
The API change LG to me.

**Comment by Yancey0623** [comment]
@typhoonzero 
Yes, all the CONTROLE MESSAGES need to be recognized in pserver.

File: paddle/fluid/framework/executor.h (hunk 1 above)
Comment (typhoonzero) [reply paddle/fluid/framework/executor.h:51]: I see

Comment (typhoonzero) [reply paddle/fluid/framework/executor.h:51]: I see

**Comment by typhoonzero** [comment]
Seems using `CHANGED_FILES` in teamcity CI cannot get overall changes of current PR (branch), so that when later commits coming in, the CI will not require approvals for api change. Not sure how to do that for now, @panyx0718 any ideas?

File: paddle/fluid/operators/distributed/grpc_client.cc (hunk 2 above)
Comment (Yancey0623) [reply paddle/fluid/operators/distributed/grpc_client.cc:40]: A trainer instance can only be completed when all completed rpc call finshed.
I think we can use a smaller scope when the pserver can recognize the controller message.

**Comment by panyx0718** [comment]
git diff with the develop latest?

**Comment by Yancey0623** [comment]
How about only diff the API.spec file? A PR may have a lot of changed files between the develop branch and PR branch if it doesn't rebase to the latest code.

**Review by panyx0718 — APPROVED** [review]
LG API change.

**Review by typhoonzero — APPROVED** [review]
LGTM

**Pull request merged**